### ISSUE TITLE
[FEATURE] Contextual text chunking processor

### ISSUE THREAD

**Issue by heemin32** [opensearch-project/neural-search#1262]
### Is your feature request related to a problem?
When we chunk the document into smaller pieces, it sometimes lose the overall context. For example, when there is a sentence like, "My dogs is small. It has white fur" and we chunk it into "My dogs is small" and "It has white fur", we lose the context and we don't know what it means here. During search, we might not be able to see correct result due to this lost context. There are few strategy to overcome this shortcoming like overlapping of texts and few mores. This contextual text chunking is another approach to overcome this issue utilizing LLM.

### What solution would you like?
We’d like to introduce a contextual text chunking processor that uses an LLM to summarize the document and then append this context to each chunk. By embedding this context, each chunk becomes more meaningful, which can improve search accuracy and relevance.

### What alternatives have you considered?
We might be able to utilize flow framework and other existing processor to mimic this behavior but haven't verified it.

### Do you have any additional context?
https://www.anthropic.com/news/contextual-retrieval

**Comment by q-andy** [comment]
I agree, this would be an exciting feature to have. Some thoughts:
- How we can integrate this with existing text chunking algorithms? Perhaps we could have this as a separate text chunking processor that can use all the same kinds of algorithms if the existing algorithms aren't tightly coupled with the existing processor.
- One purpose of text chunking is to keep text embedding/sparse model inputs below the token limit. If we configure chunk sizes to be below the token limit and then run LLMs to generate summary information, the chunk + summary info could go above the token limit again, depending on how many tokens the LLM generates. 
- Model providers can provide a max tokens parameter ([e.g. Anthropic](https://docs.anthropic.com/en/api/messages#body-max-tokens)) but since this is set in the API call users would have to configure this in the ML remote connector and keep it aligned with the processor level chunk size and their embedding generating model input size, which is confusing to configure (tracking 3 numbers in three separate places for optimal usage)
- I see many of our [pretrained models](https://docs.opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/#supported-pretrained-models) will automatically truncate when the input is above max token limit, which prevents the above case from outright failing inference, but it is not an ideal case
- We may also want to consider whether appending or prepending the chunk context would be more effective, I see both options are mentioned in the Anthropic article. This may matter in the auto-truncation case, for example, if we prepend the context at the beinning, then the chunk content will be truncated if it is above the limit. Vice versa, if we append the context at the end, the context may be truncated instead.

**Comment by q-andy** [comment]
A bottleneck that Anthropic and others call out using this strategy is the cost and performance versus regular chunking strategies. For example, say you have a text embedding model with a max token input length of 200, and you ingest a document with 3000 tokens:
- In regular chunking using a naive fixed length splitter, 3000 token document / 200 chunks = 15 total chunks = 15 total text embedding model calls
- With contextual chunking, you have 200 text embedding input tokens to allocate between the actual document chunk and the contextual summary. Say you allocate 50 tokens towards contextual summary, which means each chunk will contain 200 - 50 = 150 tokens from the document. This means you end up with 3000 / 150 = 20 chunks = 20 total text embedding model calls
- You also have the 20 LLM contextual summary calls as well for each of the chunks so there's more than twice as many API calls in total. And each will contain at least 3k tokens since you pass the entire document in the context to generate the chunk summary. You don't necessarily pay sticker price for these calls though since the 3k document length can utilize prompt caching available in many remote providers like [Bedrock](https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-caching.html), [Anthropic](https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching) and [OpenAI](https://platform.openai.com/docs/guides/prompt-caching) which reduce costs by up to 90%.

In terms of integration with OpenSearch, ingest processors can be executed on batches of documents, and different batches can be executed on in parallel to increase performance. But in this case we are making multiple LLM calls within the same processor for a single document, so we don't benefit from that existing parallelization. We'd have to implement that optimization on the processor level. Ideally we have some batching mechanism across different API calls from different ingest documents.

I'm wondering too if there's a way we could "cascade" chunks to generate embeddings for them directly after their context generation is completed, to decrease latency. If we arrange a ContextualChunker -> TextEmbeddingProcessor sequentially, we wouldn't hit the text embedding calls until all chunks are finished generating. But maybe there's performance improvement if we could pass completed chunks to the embedding generation step early, or perhaps that would be slower because we would lose performance gains from batching.

**Comment by yuye-aws** [comment]
> As `Contextual Embedding` suggested by Anthropic is most likely to be the current SOTA

By referring to SOTA, you are assuming that there is an existing benchmark on text chunking and search relevance. It would be better you can share the benchmark so that we can test the new chunking algorithm.

**Comment by yuye-aws** [comment]
> * How we can integrate this with existing text chunking algorithms? Perhaps we could have this as a separate text chunking processor that can use all the same kinds of algorithms if the existing algorithms aren't tightly coupled with the existing processor.

No need to introduce a new text chunking processor. The current chunking processor can support multiple algorithms. We just need a new algorithm class being a subclass of the **Chunker** interface

**Comment by yuye-aws** [comment]
> * One purpose of text chunking is to keep text embedding/sparse model inputs below the token limit. If we configure chunk sizes to be below the token limit and then run LLMs to generate summary information, the chunk + summary info could go above the token limit again, depending on how many tokens the LLM generates.

Thanks to @zhichao-aws . The neural-search plugin now supports two bert-based tokenizers. We can now have an estimation on the chunked tokens. Unfortunately, the vocabulary and tokenizer vary among embedding models. Some are available while others are not.

**Comment by yuye-aws** [comment]
> * I see many of our [pretrained models](https://docs.opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/#supported-pretrained-models) will automatically truncate when the input is above max token limit, which prevents the above case from outright failing inference, but it is not an ideal case

Plus one point. The bedrock titan embedding model will directly return an error if token limit exceeds. This is a tricky use case.

**Comment by yuye-aws** [comment]
> I'm wondering too if there's a way we could "cascade" chunks to generate embeddings for them directly after their context generation is completed, to decrease latency. If we arrange a ContextualChunker -> TextEmbeddingProcessor sequentially, we wouldn't hit the text embedding calls until all chunks are finished generating. But maybe there's performance improvement if we could pass completed chunks to the embedding generation step early, or perhaps that would be slower because we would lose performance gains from batching.

Good point! The ContextualChunker algorithm takes way more time than fixed token length and delimiter algorithm. That's why we need to consider stream embedding texts, but I do not think this is the current top priority.

**Comment by ramda1234786** [comment]
Hi @yuye-aws .We can go with this proposed payload and below is the example of semantic chunking.

Let's assume i got a **extract from PDF/Word document in Markdown format**. Which looks like this

```
# Employment Contract

This agreement is made between Employer and Employee on January 1, 2025.

## Compensation

The employee shall be paid a salary of $100,000 per year, payable biweekly.

## Termination Clause

Either party may terminate this agreement with a 30-day notice.
```

 **Fixed-Token Chunking Output (Bad Example — breaks mid-topic)**

```
{
  "passage_chunk": [
    "This agreement is made between Employer and Employee",
    "on January 1, 2025. The employee shall be paid",
    "a salary of $100,000 per year, payable biweekly.",
    "Either party may terminate this agreement",
    "with a 30-day notice."
  ]
}
```

**Semantic Chunking Output ( preserves structure)**

```
[
  {
    "chunk_text": "This agreement is made between Employer and Employee on January 1, 2025.",
    "metadata": {
      "section": "Introduction",
      "doc_type": "employment_contract"
    }
  },
  {
    "chunk_text": "The employee shall be paid a salary of $100,000 per year, payable biweekly.",
    "metadata": {
      "section": "Compensation",
      "doc_type": "employment_contract"
    }
  },
  {
    "chunk_text": "Either party may terminate this agreement with a 30-day notice.",
    "metadata": {
      "section": "Termination Clause",
      "doc_type": "employment_contract"
    }
  }
]
```

Add a new chunking algorithm to the existing text_chunking ingest processor that performs semantic-aware text splitting based on paragraph boundaries, sentence breaks, or Markdown headers.

The current text_chunking processor only supports fixed_token_length chunking, which is efficient but often breaks meaningful text structures, such as:

- Mid-sentence
- Mid-paragraph
- Between headers and body text in Markdown

This leads to loss of semantic context, which negatively affects the performance of downstream vector search, retrieval-augmented generation (RAG), and LLM-powered applications that depend on high-quality contextual input.

A semantic chunking mode would improve:

- Chunk-level coherence
- Embedding quality
- Search relevance
- Prompt reconstruction accuracy

Introduce a new algorithm type under text_chunking called "semantic"

```
PUT _ingest/pipeline/text-semantic-chunking
{
  "description": "Semantic-aware text chunking pipeline with block preservation",
  "processors": [
    {
      "text_chunking": {
        "algorithm": {
          "semantic": {
            "mode": "markdown",              // or "paragraph", "sentence"
            "chunk_size_hint": 500,          // (optional) soft token/character length per chunk
            "overlap_rate": 0.2,             // (optional) overlap between chunks
            "preserve_blocks": [             // (optional) preserve structural blocks in markdown
              "table",                       // do not split markdown tables
              "list",                        // preserve ordered/unordered lists
              "code"                         // preserve fenced code blocks
            ]
          }
        },
        "field_map": {
          "passage_text": "semantic_chunks"
        }
      }
    }
  ]
}

```

**Supported Modes**

| Mode        | Behavior                                                                            |
|-------------|-------------------------------------------------------------------------------------|
| `markdown`  | Splits by Markdown structure (e.g., `#`, `##`, `###`) while grouping nearby content |
| `paragraph` | Splits on double line breaks (`\n\n`) or punctuation followed by newline            |
| `sentence`  | Uses sentence segmentation (e.g., via pre-trained model or heuristic)               |


With this you are achieving:-

- Improved retrieval relevance for RAG and vector search
- Preserves document semantics (contracts, legal docs, reports)
- Reduces noise in chunk embeddings
- Aligns with chunking standards from LangChain, Haystack, and LlamaIndex

I hope with this approach and solution, it will be a rockstar chunker as compared to LangChain, Haystack, and LlamaIndex

**Comment by yuye-aws** [comment]
@ramda1234786 For your supported modes. You can check the delimiter algorithm. You can cascade a few delimiter processors.
